cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,484 @@
+	if data == nil {
+		return nil, false
+	}
+	if nodeInfo, found := data.addedNodeInfoMap[name]; found {
+		return nodeInfo, true
+	}
+	if nodeInfo, found := data.modifiedNodeInfoMap[name]; found {
+		return nodeInfo, true
+	}
+	return nil, false
+}
+
+func (data *internalDeltaSnapshotData) getNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	if data == nil {
+		return nil
+	}
+	if data.nodeInfoList == nil {
+		data.nodeInfoList = data.buildNodeInfoList()
+	}
+	return data.nodeInfoList
+}
+
+// Contains costly copying throughout the struct chain. Use wisely.
+func (data *internalDeltaSnapshotData) buildNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	baseList := data.baseData.getNodeInfoList()
+	totalLen := len(baseList) + len(data.addedNodeInfoMap)
+	var nodeInfoList []*schedulernodeinfo.NodeInfo
+
+	if len(data.deletedNodeInfos) > 0 || len(data.modifiedNodeInfoMap) > 0 {
+		nodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, totalLen+100)
+		for _, bni := range baseList {
+			if data.deletedNodeInfos[bni.Node().Name] {
+				continue
+			}
+			if mni, found := data.modifiedNodeInfoMap[bni.Node().Name]; found {
+				nodeInfoList = append(nodeInfoList, mni)
+				continue
+			}
+			nodeInfoList = append(nodeInfoList, bni)
+		}
+	} else {
+		nodeInfoList = make([]*schedulernodeinfo.NodeInfo, len(baseList), totalLen+100)
+		copy(nodeInfoList, baseList)
+	}
+
+	for _, ani := range data.addedNodeInfoMap {
+		nodeInfoList = append(nodeInfoList, ani)
+	}
+
+	return nodeInfoList
+}
+
+// Convenience method to avoid writing loop for adding nodes.
+func (data *internalDeltaSnapshotData) addNodes(nodes []*apiv1.Node) error {
+	for _, node := range nodes {
+		if err := data.addNode(node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addNode(node *apiv1.Node) error {
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	if err := nodeInfo.SetNode(node); err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo: %v", err)
+	}
+	return data.addNodeInfo(nodeInfo)
+}
+
+func (data *internalDeltaSnapshotData) addNodeInfo(nodeInfo *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.getNodeInfo(nodeInfo.Node().Name); found {
+		return fmt.Errorf("node %s already in snapshot", nodeInfo.Node().Name)
+	}
+	data.addedNodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.addedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+	if _, found := data.modifiedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
```

Comment:
If this node exists in base (which it does if we reach this call), you'll add it to removedNI inside removeNode(). In follow-up addNodeInfo you will add it back to modifiedNIMap, but you won't clean it from removedNI leading to inconsistent internal state.
Incidentally you only ever call updateNode() on base data in commit(). You're probably better of getting rid of this method altogether, rather than trying to make it correct.